New renaming policy

craff opened the pull request:
This pull request has the following effects:
- only one way to print term with renaming of variables to avoid clash: using context
- doc updated to say that we recommand to postpone renaming to printing if possible
- a functorial interface to ctxt is provided to customize renaming solving issue #11 
- performance should be improved on pure computation (without printing)

This PR should be tested agains some significative code like pml,subml,lambdapi

This PR breaks a lot of existing code that do not use context for printing (it only breaks printing though)

craff committed:
Require the use of contexts for printing.

THIS IS A BREAKING CHANGE.

No more renaming is done at term construction (if a context is not
used): this only needs to happen at printing time.

A notion of renaming policy is added to control how fresh variable
names are constructed, and how variable names are printed.

Documentation has been adapted and cleaned up. Duplicated comments
have been removed from the [.ml] file.

Other changes:
- New benchmarks (use [make bench]).
- The opam file is generated with dune.
- A CHANGELOG.md has been created.
- We require OCaml >= 4.07.0 to avoid CI annoyance.

Co-authored-by: Rodolphe Lepigre <lepigre@mpi-sws.org>

rlepigre commented:
If we decide to enforce the use of contexts for renaming then we should probably get rid of the functions that do not work under a context, don't you think? It just makes no sense to print terms for which names may appear to clash. That should not be allowed by the library, and we should enforce the use of contexts whenever they are necessary to avoid clashes.

(I have not looked in detail at the PR, I'll try to find some time soon but I am super with an approaching deadline.)

rlepigre commented:
@craff I also rebased this branch on master, so that we have CI with a GitHub workflow. Currently the tests fail because the output of some examples is different from what we had before. I have not looked at this in details yet, but if you want to try you can run `make tests` to see the diff between the test result and what was expected. If the commands return the expected result but the `.expected` files are not up to date you can run `make promote` to generate them from the command output.

craff commented:
Normal that the output changes. 

Renaming by copying allow lx ly.ly to be rename as lx lx.lx. With context, x and y are in the context and distinct name are enforced.

We can create a renaming policy that check the variable really occuring to simplify the context. Another good option is to reset the context at least if the term is closed. I will change this.

craff commented:
I have added a way to control when renaming is done or not. You can even use "_" for binder that do not bind anything.
See lambda.ml and documentation ...

craff commented:
Regarding checking when name_of is safe (that is when printing is safe):
In the master branch, this is already done and the rule for correct printing is:
- either use context
- or copy the term with a lift like function prior to printing

In this PR
- you must use ctxt

No check was performed before and the usage rule is simpler.

moreover, using name_of directly might be very usefull when debugging, even if names are not correct.

craff commented:
The remark in issue #11 about zeros is fixed by the patch above: x0 is renamed in x01, x02 etc and the 0 is not erased anymore.

craff commented:
After a quick test: this PR seems to improve time efficiency by 20% and a bit of memory gain too.

craff commented:
Remark: it is possible to write a lift function for terms that uses a context and renames all variables. Therefore correct printing can be achieved in two separate steps: renaming, then printing. 

Moreover, normalisation can also use a ctxt and rename terms if one wants. This add useless term traversals, that normalisation does not need in general (depending on the normalisation strategy). Therefore this is not a very good idea. Actually, this is why normalisation could leave wrong names even in the previous version: extra term traversal where also necessary just for renaming.

rlepigre commented:
@craff I also fixed the branch here to be based on master. Please, make sure that working branches are always based on master so that we don't have annoying merge commits there.

Do do that, you need to do something like `git rebase master` instead of a merge.

craff commented:
> If we decide to enforce the use of contexts for renaming then we should probably get rid of the functions that do not work under a context, don't you think? 

No, because if you do normalisation, conversion or other function which are not intended to print, you do not want to do renaming.

rlepigre commented:
I pushed a bit of documentation cleanup, but I'm not done. I'll continue tomorrow.

craff commented:
I worked a bit more on the documentation and merged master which not as the existantial patch.

craff commented:
J'ai essayé de faire un git rebase master ... ça fait des conflits, je ne comprends rien.
Argh, je comprends, comme j'ai fait un merge après ton rebase, et avant de lire ton message, c'est la mouise
pour refaire un rebase. Bon je ferais plus que des rebase ... après le merge final de cette PR ? 
C'est le plus simple ?

rlepigre commented:
Bon, j'ai réparé la branche une dernière fois. Il faudra que tu fasses un `reset` de ta branche.
Si tu n'as fait aucun travail en plus, tu peux faire `git fetch` puis (dans ta branche locale) `git reset --hard origin/new_renaming_policy`.

craff commented:
J'ai pas touché... Et je ne fais plus de merge. Que des rebase sur master ou head~n avec la bonne valeur de n. Merci. 

Le 24 janvier 2022 22:59:24 GMT-10:00, Rodolphe Lepigre ***@***.***> a écrit :
>Bon, j'ai réparé la branche une dernière fois. Il faudra que tu fasses un `reset` de ta branche.
>
>Si tu n'as fait aucun travail en plus, tu peux faire `git fetch` puis (dans ta branche locale) `git reset --hard origin/new_renaming_policy`.
>
>
>
>-- >
>Reply to this email directly or view it on GitHub:
>
>https://github.com/rlepigre/ocaml-bindlib/pull/12#issuecomment-1020950061
>
>You are receiving this because you were mentioned.
>
>
>
>Message ID: ***@***.***>

craff commented:
There is a strange error when compiling with 4.02: it says dune require 4.08, but compilation with 4.03 works ... I do no get it ?

craff commented:
could reproduce on my machine. opam install dune before opam upgrade works (dune require to install ocaml >= 4.08 for installation). I don't understand why it fails, probably a temporary bug in opam ?

rlepigre commented on .github/workflows/ci.yml:
```diff
@@ -30,6 +30,7 @@ jobs:
         run: |
           eval $(opam env)
           opam update
+          opam install dune
```
That is really weird: why do we need that? It seems to have fixed the problem but the `opam install --deps-only -d -t bindlib` should take care of that.

craff replied:
I think this is a temporary bug in opam packages ? the installation of dune via ocaml 4.08 for older version of ocaml
was failing at opam upgrade but opam install dune works ...I think it is harmless anyway ? We can recheck in a few days 
if opam is updated ? Or open an issue on opam ?

craff replied:
I leave the conversation unresolved ... We will decide before you are ready to merge. Maybe everything will be fixed by that time ?

rlepigre replied:
OK, let's see.

craff commented:
@rlepigre On pourrait mettre un

include module type of Default_Ctxt dans bindlib.mli
pour n'avoir qu'une fois les fonctions XXX_in.

Par contre, elle serait dans le foncteur uniquement.

Tu en penses quoi ?

craff commented:
module type of is available since 3.12

rlepigre commented:
> Par contre, elle serait dans le foncteur uniquement.
> 
> Tu en penses quoi ?

J'aime pas trop cette idée, je pense que ça complique l'interface. C'est pas super grave s'il y a un peu de duplication, surtout que dans le foncteur on peut just pointer vers la doc des fonctions définies à l'extérieur.

rlepigre commented:
@craff est-ce que ça te pose un problème si je vire le type `renaming_policy` et que j'inline ses champs dans le paramètre du foncteur? Je trouve que c'est pas utile d'avoir à la fois un record et un foncteur, on pourrait avoir l'un ou l'autre (donc plutôt le foncteur).

craff commented:
> > Par contre, elle serait dans le foncteur uniquement.
> > Tu en penses quoi ?
> 
> J'aime pas trop cette idée, je pense que ça complique l'interface. C'est pas super grave s'il y a un peu de duplication, surtout que dans le foncteur on peut just pointer vers la doc des fonctions définies à l'extérieur.

Je suis en fait d'accord, mais je voulais te poser la question.

craff commented:
> @craff est-ce que ça te pose un problème si je vire le type `renaming_policy` et que j'inline ses champs dans le paramètre du foncteur? Je trouve que c'est pas utile d'avoir à la fois un record et un foncteur, on pourrait avoir l'un ou l'autre (donc plutôt le foncteur).

ça me va très bien. Par contre, pull, je viens de corriger un vrai bug dans le foncteur.

craff commented:
Sinon je dois encore faire une connerie. Regarde le log de test_ocaml_version2 (avec du sed) tout mes rebase + push
pour suivre le branche new_renaming_policy me font un merge et une duplication de commit bizarre. Il me manque une option à rebase ou push ?

rlepigre commented:
> Sinon je dois encore faire une connerie. Regarde le log de test_ocaml_version2 (avec du sed) tout mes rebase + push pour suivre le branche new_renaming_policy me font un merge et une duplication de commit bizarre. Il me manque une option à rebase ou push ?

Je sais pas quel est le problème, mais à mon avis c'est pas la peine de te faire chier à synchroniser les branches. On fera un rebase de ton truc avec `sed` quand on aura mergé cette PR. En pratique, c'est rarement une bonne idée de rebaser sur une branche de travail, et chaque PR reste basée sur `master`. Ensuite quand tu en merge une, tu rebase l'autre sur `master`, et il n'y a pas de problème.

rlepigre commented on lib/bindlib.ml:
```diff
@@ -664,321 +661,238 @@ let bind_mvar : 'a mvar -> 'b box -> ('a,'b) mbinder box = fun xs b ->
-  let get_suffix name suffix ctxt =
-    try
-      let n = SMap.find name ctxt in
-      if suffix > n then (suffix, ctxt) else
```
Je crois qu'il y avait un bug ici: le contexte n'était pas mis à jour. Tu confirmes @craff? Normalement c'est réparé par mon dernier commit. D'ailleurs, j'ai ajouté un peu de documentation, et enlevé les commentaires de documentation sur certaines fonctions. (C'est dommage, mais c'est trop compliqué de maintenir les commentaires à jour entre le `.ml` et le `.mli`. J'ai aussi viré quelques types.)

craff replied:
Tu as raison. Tu as pushé ton commit ? Sinon, j'ai corrigé en local, mais j'ai pas pushé ...

craff replied:
C'est bon, j'ai trouvé ton commit. C'est parfait.

rlepigre reviewed (approved):
I think this looks pretty good now.

rlepigre commented:
Avant de merger, ce serait bien de créer un `CHANGELOG.md` pour explique les différences avec la version précédente, et comment faire pour porter son code. (D'ailleurs, j'ai changé le tag de version dans les sources.)

rlepigre commented:
@craff there is some issue with the generation of the documentation using `odoc` on older versions of OCaml. I'm getting tired of this kind of stuff, so I propose we just drop support for all versions `<= 4.07.0`. I don't think anyone uses such old versions of OCaml, and if they do they are unlikely to want to use Bindlib.

rlepigre commented on CHANGELOG.md:
```diff
@@ -0,0 +1,163 @@
+This releases breaks your printing code: here is how to fix it
+--------------------------------------------------------------
+
+TODO
```
We still need to write a guide on how to port code to this new version of Bindlib.

rlepigre replied:
C'est fait, tu peux jeter un œil @craff?

rlepigre commented:
@craff une question que je me pose: est-ce que tu crois qu'on devrait fournir plus d'opérations sur le type `ctxt`? (Par exemple une fonction `union`.)

rlepigre commented:
@craff une autre question: les contextes par défaut utilisent le type `int SMap.t`, mais est-ce qu'on ne devrait pas carrément passer à `ISet SMap.t` (donc remplacer l'entier par un ensemble d'entiers), et renommer en utilisant le premier suffixe disponible?

Maintenant qu'on fait le renommage juste avant l'affichage, on peut sans problème se permettre d'utiliser des opérations moins efficaces comme elles seront toujours négligeables à côté des entrées-sorties du programme.

craff commented:
On 22-02-07 13:26:47, Rodolphe Lepigre wrote:
> @craff une autre question: les contextes par défaut utilisent le type int
> SMap.t, mais est-ce qu'on ne devrait pas carrément passer à ISet SMap.t (donc
> remplacer l'entier par un ensemble d'entiers), et renommer en utilisant le
> premier suffixe disponible?
>
Tu as raison!

C'est assez chiant le code qui cherche le premier entier qui n'est pas dans un ensemble.
Certe les ensembles sont petits en pratique, mais j'aime pas trop par principe.

Idéalement, il faudrait une bonne représentation des ensembles d'entiers comme
réunion d'intervalles, que l'on place dans un arbre bien balancé.
J'ai ça quelque-part ... ça donne des opérations
en O(ln(n)) et évite les surprises si quelqu'un fait une variable x10000000.

Mais bon l'implémantation juste avec des ensembles d'entiers ça peut aller.

Faut quand même rester avec une compléxité raisonnable, car les algos de
typage se traîne un contexte ... pour afficher les messages d'erreurs ou pour
un mode verbeux ou pour l'affichage des buts dans le cas de pml2.

A+
Christophe


> Maintenant qu'on fait le renommage juste avant l'affichage, on peut sans
> problème se permettre d'utiliser des opérations moins efficaces comme elles
> seront toujours négligeables à côté des entrées-sorties du programme.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.
> You are receiving this because you were mentioned.*Message ID: <rlepigre/
> ***@***.***>
>

--
Christophe Raffalli
tél: +689 87 23 11 48
web: http://raffalli.eu

craff commented:
On 22-02-07 13:18:04, Rodolphe Lepigre wrote:
> @craff une question que je me pose: est-ce que tu crois qu'on devrait fournir
> plus d'opérations sur le type ctxt? (Par exemple une fonction union.)

Je ne vois pas trop de cas où on aurait besoin de faire ça.

On attends des demandes ?

> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.
> You are receiving this because you were mentioned.*Message ID: <rlepigre/
> ***@***.***>
>

--
Christophe Raffalli
tél: +689 87 23 11 48
web: http://raffalli.eu

rlepigre commented:
> Idéalement, il faudrait une bonne représentation des ensembles d'entiers comme réunion d'intervalles, que l'on place dans un arbre bien balancé. J'ai ça quelque-part ... ça donne des opérations en O(ln(n)) et évite les surprises si quelqu'un fait une variable x10000000. Mais bon l'implémantation juste avec des ensembles d'entiers ça peut aller.

OK, j'ai implanté ce truc rapidement, mais je ne le push pas ici. Je pense que ce serait bien de merger ce qu'on a déjà, et je ferai une nouvelle PR pour le changement de représentation du contexte.

Par contre, avant de merger, il faudrait qu'on écrive un bout de doc (TODO dans `CHANGELOG.md`) qui explique comment porter le code de Bindlib 5.0.1 à Bindlib 6.0. A priori c'est asse simple en utilisant les fonctions que j'ai ajouté: on peut reconstruire un contexte initial (contenant toutes les variables libres) à partir d'un terme boxé. Donc ça veut dire qu'il faut faire les changements suivants dans une fonction d'affichage pour un terme `t`:
- boxer `t` et utiliser `free_vars (box_term t)` pour obtenir un context de départ,
- passer le context comme argument de la fonction d'affichage, et utiliser `unbind_in` et compagnie.

Une alternative serait de suggérer d'implanter une fonction `cleanup`, qui avant était juste `fun t -> unbox (box_term t)`, mais qui deviendrait une fonction récursive similaire à `box_term`, mais qui utilise un contexte. En gros, ça veut dire qu'il est toujours correct d'afficher `cleanup t` de manière naive.

Tu en penses quoi?

craff commented:
On 22-02-11 16:37:55, Rodolphe Lepigre wrote:
>     Idéalement, il faudrait une bonne représentation des ensembles d'entiers
>     comme réunion d'intervalles, que l'on place dans un arbre bien balancé.
>     J'ai ça quelque-part ... ça donne des opérations en O(ln(n)) et évite les
>     surprises si quelqu'un fait une variable x10000000. Mais bon
>     l'implémantation juste avec des ensembles d'entiers ça peut aller.
>
> OK, j'ai implanté ce truc rapidement, mais je ne le push pas ici. Je pense que
> ce serait bien de merger ce qu'on a déjà, et je ferai une nouvelle PR pour le
> changement de représentation du contexte.

OK.

> Par contre, avant de merger, il faudrait qu'on écrive un bout de doc (TODO dans
> CHANGELOG.md) qui explique comment porter le code de Bindlib 5.0.1 à Bindlib
> 6.0. A priori c'est asse simple en utilisant les fonctions que j'ai ajouté: on
> peut reconstruire un contexte initial (contenant toutes les variables libres) à
> partir d'un terme boxé. Donc ça veut dire qu'il faut faire les changements
> suivants dans une fonction d'affichage pour un terme t:
>
>   • boxer t et utiliser free_vars (box_term t) pour obtenir un context de
>     départ,
>   • passer le context comme argument de la fonction d'affichage, et utiliser
>     unbind_in et compagnie.

Oui.

> Une alternative serait de suggérer d'implanter une fonction cleanup, qui avant
> était juste fun t -> unbox (box_term t), mais qui deviendrait une fonction
> récursive similaire à box_term, mais qui utilise un contexte. En gros, ça veut
> dire qu'il est toujours correct d'afficher cleanup t de manière naive.

il fallait déjà que box_term soit une fonction récursive, traversant tout le terme.

Je ne sais pas si on veut parler de cette solution. Dans beaucoup de contexte (compilation,
typage, etc ...) on veut parcourir le terme avec un contexte pour afficher les messages
d'erreurs ou de débuggage. Et souvent on a déjà une notion de contexte lors de ce parcourt.

Donc la meilleur suggestion pour moi c'est de dire:

"""
Functions that scan recursively a term (like typing a term) and that may need
to print some sub-terms (for an error message or some debugging information)
should preferably use a context an the function that updates it (like unbind_in, newvar_in, ...).
This applies to the printing function itself
"""

Je ne sais pas si je veux parler de la solution de parcourt et renommage des termes.
Si on veut afficher un sous terme lors d'un parcourt, ça ne marchera pas.

> Tu en penses quoi?
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.
> You are receiving this because you were mentioned.*Message ID: <rlepigre/
> ***@***.***>
>

--
Christophe Raffalli
tél: +689 87 23 11 48
web: http://raffalli.eu

rlepigre commented:
> Je ne sais pas si on veut parler de cette solution. Dans beaucoup de contexte (compilation,
typage, etc ...) on veut parcourir le terme avec un contexte pour afficher les messages
d'erreurs ou de débuggage. Et souvent on a déjà une notion de contexte lors de ce parcourt.

C'est vrai, mais ça veut aussi dire que pour vérifier le typage dans un language où il y a des types dépendants, on paye le coût de la gestion du contexte même quand on ne veut pas faire d'affichage du tout, ce qui est peut-être dommage? Mon idée c'était que dés lors que tu veut afficher des messages pour débugger, tu paye déjà le coût de `printf` (ou similaire), et qu'un parcours linéaire (ou un peu plus) de terme ne fait pas grande différence.

rlepigre committed:
Delete the [attic] directory.

This includes:
- Porting old [unif.ml] example.
- Porting old [pred2.ml] example.
- Porting old [metavar] example.
- Integrating examples from the old doc.
- Creating [SEMANTICS.md] with interesting parts of the old doc.
- Porting old benchmarks (de Bruijn + Bindoib).

rlepigre commented:
I squashed a few commits, I'll merge now. There are still a few parts (mostly documentation and examples / benchmarks) that would benefit from a review @craff. I guess you can just look over things when you have a moment, and create an issue if you find problems.
